### PR TITLE
mds/snapdiff: snapdiff doesn't report modified files

### PR DESCRIPTION
**Pull request by kotreshhr**
The snapdiff doesn't report files which are modified in following case
1. If the modified files between two snapshots has snap flush pending from clients
    The following upstream teuthology tests caught this.
        a. test_cephfs_mirror_incremental_sync
        b. test_cephfs_mirror_sync_with_purged_snapshot
 

The PR takes care of fixing them.

Fixes: https://tracker.ceph.com/issues/74984

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)

You must only issue one Jenkins command per-comment. Jenkins does not understand
comments with more than one command.
</details>

### CODE REVIEW

File: src/mds/Server.cc
Comment (kotreshhr) [inline src/mds/Server.cc:12628]: @vshankar 
The files are skipped by snapdiff in three conditions
1. The two snapshots are within in dentry's first and last
2. The two snapshots are outside dentry's first and last
3. inode and mtime comparison

Taking file lock on every dentry for snap flush before three conditions had negative effect on performance. It's not feasible. So I optimised that as below.
1. Take the file lock only if mtime comparison matches below.
2. Take the file lock only on the latest snapshot's dentry (it's also head, if dentry isn't split)

With above changes, the upstream tests which were failing passes - https://pulpito-ng.ceph.com/runs/khiremat-2026-06-29_15:33:15-fs-wip-khiremat-snapdiff-snapflush-bug-29062026-distro-default-trial

But my question is, can there be scenarios where the dentry split has not happened (case 1) because of pending snap flush that can cause file to be missed?

Comment (kotreshhr) [reply src/mds/Server.cc:12628]: @vshankar ptal, I have used pending snap flush condition as well.

Comment (vshankar) [reply src/mds/Server.cc:12628]: Sure.

**Comment by vshankar** [comment]
PR https://github.com/ceph/ceph/pull/68054 adds functionality to specify mask for file attribute changes. I haven't reviewed it yet, but it seems like the changes introduced in that PR to specify mask params is probably a bit more flexible.

File: src/mds/Server.cc [hunk 1]
```diff
@@ -12551,8 +12584,40 @@ bool Server::build_snap_diff(
       }
     } else {
       if (snapid_prev >= dn->first && snapid <= dn->last) {
-	dout(20) << __func__ << " skipping unchanged " << dn->get_name() << " "
-	  << dn->first << "/" << dn->last << dendl;
+	if (!snapflush_pending(in)) {
+	  dout(20) << __func__ << " skipping unchanged " << dn->get_name() << " "
+	    << dn->first << "/" << dn->last << dendl;
+	  continue;
```
[the remark this answers is not in the document]
Comment (kotreshhr) [reply src/mds/Server.cc:12632]: preserving mtime issue is fixed as part of separate PR

**Comment by kotreshhr** [comment]
Sample run - https://pulpito-ng.ceph.com/runs/khiremat-2026-08-11_11:02:21-fs-wip-khiremat-snapdiff-snapflush-bug-11082026-distro-default-trial

The snapdiff duration with the fix. Observe increment_duration2
```
khiremat@trial200:/teuthology/khiremat-2026-08-11_11:02:21-fs-wip-khiremat-snapdiff-snapflush-bug-11082026-distro-default-trial$ grep "full_sync_duration" */teuthology.log
456477/teuthology.log:2026-08-11T13:32:48.160 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 12.0, inc_sync_duration1 - 6.0
456477/teuthology.log:2026-08-11T13:33:24.879 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 12.0, inc_sync_duration2 - 21.0
456478/teuthology.log:2026-08-11T13:32:24.254 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 14.0, inc_sync_duration1 - 6.0
456478/teuthology.log:2026-08-11T13:33:00.105 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 14.0, inc_sync_duration2 - 17.0
456479/teuthology.log:2026-08-11T13:32:46.386 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 13.0, inc_sync_duration1 - 6.0
456479/teuthology.log:2026-08-11T13:33:23.096 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 13.0, inc_sync_duration2 - 21.0
456480/teuthology.log:2026-08-11T13:32:48.924 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 16.0, inc_sync_duration1 - 5.0
456480/teuthology.log:2026-08-11T13:33:25.017 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 16.0, inc_sync_duration2 - 18.0
456481/teuthology.log:2026-08-11T13:32:25.668 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 15.0, inc_sync_duration1 - 9.0
456481/teuthology.log:2026-08-11T13:33:01.993 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 15.0, inc_sync_duration2 - 18.0
456487/teuthology.log:2026-08-11T13:47:37.221 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 22.0, inc_sync_duration1 - 11.0
456487/teuthology.log:2026-08-11T13:48:17.467 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 22.0, inc_sync_duration2 - 22.0
456488/teuthology.log:2026-08-11T13:49:00.471 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 23.0, inc_sync_duration1 - 13.0
456488/teuthology.log:2026-08-11T13:49:40.099 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 23.0, inc_sync_duration2 - 24.0
456489/teuthology.log:2026-08-11T13:48:02.538 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 22.0, inc_sync_duration1 - 11.0
456489/teuthology.log:2026-08-11T13:48:42.248 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 22.0, inc_sync_duration2 - 22.0
456490/teuthology.log:2026-08-11T13:47:19.346 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 22.0, inc_sync_duration1 - 11.0
456490/teuthology.log:2026-08-11T13:47:58.913 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 22.0, inc_sync_duration2 - 22.0
456491/teuthology.log:2026-08-11T13:48:13.076 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 21.0, inc_sync_duration1 - 10.0
456491/teuthology.log:2026-08-11T13:48:52.602 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 21.0, inc_sync_duration2 - 21.0
```

Comparing the same with the sample tentacle run without multithreaded mirroring - https://pulpito-ng.ceph.com/runs/khiremat-2026-08-11_15:29:17-fs-wip-khiremat-mirror-backports-tentacle-10082026-distro-default-trial
Note that I had the this log change 

```
khiremat@trial200:/teuthology/khiremat-2026-08-11_15:29:17-fs-wip-khiremat-mirror-backports-tentacle-10082026-distro-default-trial$ grep "full_sync_duration" */teuthology.log
457492/teuthology.log:2026-08-11T16:06:29.098 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 19.0, inc_sync_duration1 - 16.0
457492/teuthology.log:2026-08-11T16:07:14.827 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 19.0, inc_sync_duration2 - 14.0
457493/teuthology.log:2026-08-11T16:06:22.349 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 17.0, inc_sync_duration1 - 11.0
457493/teuthology.log:2026-08-11T16:07:18.448 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 17.0, inc_sync_duration2 - 13.0
457494/teuthology.log:2026-08-11T16:07:01.481 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 20.0, inc_sync_duration1 - 16.0
457494/teuthology.log:2026-08-11T16:07:47.193 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 20.0, inc_sync_duration2 - 17.0
457495/teuthology.log:2026-08-11T16:07:09.866 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 17.0, inc_sync_duration1 - 6.0
457495/teuthology.log:2026-08-11T16:08:05.575 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 17.0, inc_sync_duration2 - 16.0
457496/teuthology.log:2026-08-11T16:06:34.190 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 18.0, inc_sync_duration1 - 6.0
457496/teuthology.log:2026-08-11T16:07:30.144 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 18.0, inc_sync_duration2 - 16.0
457502/teuthology.log:2026-08-11T16:16:46.770 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 22.0, inc_sync_duration1 - 11.0
457502/teuthology.log:2026-08-11T16:17:26.428 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 22.0, inc_sync_duration2 - 21.0
457503/teuthology.log:2026-08-11T16:17:02.551 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 22.0, inc_sync_duration1 - 11.0
457503/teuthology.log:2026-08-11T16:17:42.212 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 22.0, inc_sync_duration2 - 21.0
457504/teuthology.log:2026-08-11T16:16:56.626 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 23.0, inc_sync_duration1 - 9.0
457504/teuthology.log:2026-08-11T16:17:36.596 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 23.0, inc_sync_duration2 - 21.0
457505/teuthology.log:2026-08-11T16:16:57.587 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 22.0, inc_sync_duration1 - 9.0
457505/teuthology.log:2026-08-11T16:17:37.341 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 22.0, inc_sync_duration2 - 20.0
457506/teuthology.log:2026-08-11T16:16:09.330 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 23.0, inc_sync_duration1 - 11.0
457506/teuthology.log:2026-08-11T16:16:48.693 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 23.0, inc_sync_duration2 - 21.0
```

File: src/mds/Server.cc
Comment (kotreshhr) [reply src/mds/Server.cc:12628]: @vshankar As discussed over the meeting, I spent time on dentry split for the following scenario.

1. snap1 (mksnap dir1)  → COW dir1's parent dentry only  (say, file1 dentry, [2, HEAD]
2. buffered write       → MDS unaware; file1 dentry unchanged
3. snap2 (mksnap dir1)  → COW dir1's parent dentry only
                                     → file1 dentry STILL [2, HEAD]
                                     
 Then dentry split happens in `journal_cow_dentry` which gets called in following scenarios asynchronously.
 Note that the `mksnap` does call `journal_cow_dentry` but on the parent directory inode. It doesn't walk the children. 
 
 1. When snap flush which is queued as part of insert_trace is executed
 2. Cap updates (_do_cap_update)
 3. Metadata ops which indirectly calls `journal_cow_dentry` like setattr, unlink....
 
 So the `snapflush_pending` is very much required in the above scenario and file1 dentry would be still [2, HEAD]

**Comment by vshankar** [comment]
> Sample run - https://pulpito-ng.ceph.com/runs/khiremat-2026-08-11_11:02:21-fs-wip-khiremat-snapdiff-snapflush-bug-11082026-distro-default-trial
> 
> The snapdiff duration with the fix. Observe increment_duration2
> 
> ```
> khiremat@trial200:/teuthology/khiremat-2026-08-11_11:02:21-fs-wip-khiremat-snapdiff-snapflush-bug-11082026-distro-default-trial$ grep "full_sync_duration" */teuthology.log
> 456477/teuthology.log:2026-08-11T13:32:48.160 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 12.0, inc_sync_duration1 - 6.0
> 456477/teuthology.log:2026-08-11T13:33:24.879 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 12.0, inc_sync_duration2 - 21.0
> 456478/teuthology.log:2026-08-11T13:32:24.254 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 14.0, inc_sync_duration1 - 6.0
> 456478/teuthology.log:2026-08-11T13:33:00.105 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 14.0, inc_sync_duration2 - 17.0
> 456479/teuthology.log:2026-08-11T13:32:46.386 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 13.0, inc_sync_duration1 - 6.0
> 456479/teuthology.log:2026-08-11T13:33:23.096 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 13.0, inc_sync_duration2 - 21.0
> 456480/teuthology.log:2026-08-11T13:32:48.924 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 16.0, inc_sync_duration1 - 5.0
> 456480/teuthology.log:2026-08-11T13:33:25.017 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 16.0, inc_sync_duration2 - 18.0
> 456481/teuthology.log:2026-08-11T13:32:25.668 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 15.0, inc_sync_duration1 - 9.0
> 456481/teuthology.log:2026-08-11T13:33:01.993 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 15.0, inc_sync_duration2 - 18.0
> 456487/teuthology.log:2026-08-11T13:47:37.221 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 22.0, inc_sync_duration1 - 11.0
> 456487/teuthology.log:2026-08-11T13:48:17.467 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 22.0, inc_sync_duration2 - 22.0
> 456488/teuthology.log:2026-08-11T13:49:00.471 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 23.0, inc_sync_duration1 - 13.0
> 456488/teuthology.log:2026-08-11T13:49:40.099 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 23.0, inc_sync_duration2 - 24.0
> 456489/teuthology.log:2026-08-11T13:48:02.538 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 22.0, inc_sync_duration1 - 11.0
> 456489/teuthology.log:2026-08-11T13:48:42.248 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 22.0, inc_sync_duration2 - 22.0
> 456490/teuthology.log:2026-08-11T13:47:19.346 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 22.0, inc_sync_duration1 - 11.0
> 456490/teuthology.log:2026-08-11T13:47:58.913 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 22.0, inc_sync_duration2 - 22.0
> 456491/teuthology.log:2026-08-11T13:48:13.076 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 21.0, inc_sync_duration1 - 10.0
> 456491/teuthology.log:2026-08-11T13:48:52.602 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 21.0, inc_sync_duration2 - 21.0
> ```
> 
> Comparing the same with the sample tentacle run without multithreaded mirroring - https://pulpito-ng.ceph.com/runs/khiremat-2026-08-11_15:29:17-fs-wip-khiremat-mirror-backports-tentacle-10082026-distro-default-trial Note that I had the this log change
> 
> ```
> khiremat@trial200:/teuthology/khiremat-2026-08-11_15:29:17-fs-wip-khiremat-mirror-backports-tentacle-10082026-distro-default-trial$ grep "full_sync_duration" */teuthology.log
> 457492/teuthology.log:2026-08-11T16:06:29.098 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 19.0, inc_sync_duration1 - 16.0
> 457492/teuthology.log:2026-08-11T16:07:14.827 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 19.0, inc_sync_duration2 - 14.0
> 457493/teuthology.log:2026-08-11T16:06:22.349 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 17.0, inc_sync_duration1 - 11.0
> 457493/teuthology.log:2026-08-11T16:07:18.448 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 17.0, inc_sync_duration2 - 13.0
> 457494/teuthology.log:2026-08-11T16:07:01.481 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 20.0, inc_sync_duration1 - 16.0
> 457494/teuthology.log:2026-08-11T16:07:47.193 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 20.0, inc_sync_duration2 - 17.0
> 457495/teuthology.log:2026-08-11T16:07:09.866 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 17.0, inc_sync_duration1 - 6.0
> 457495/teuthology.log:2026-08-11T16:08:05.575 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 17.0, inc_sync_duration2 - 16.0
> 457496/teuthology.log:2026-08-11T16:06:34.190 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 18.0, inc_sync_duration1 - 6.0
> 457496/teuthology.log:2026-08-11T16:07:30.144 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 18.0, inc_sync_duration2 - 16.0
> 457502/teuthology.log:2026-08-11T16:16:46.770 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 22.0, inc_sync_duration1 - 11.0
> 457502/teuthology.log:2026-08-11T16:17:26.428 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 22.0, inc_sync_duration2 - 21.0
> 457503/teuthology.log:2026-08-11T16:17:02.551 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 22.0, inc_sync_duration1 - 11.0
> 457503/teuthology.log:2026-08-11T16:17:42.212 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 22.0, inc_sync_duration2 - 21.0
> 457504/teuthology.log:2026-08-11T16:16:56.626 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 23.0, inc_sync_duration1 - 9.0
> 457504/teuthology.log:2026-08-11T16:17:36.596 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 23.0, inc_sync_duration2 - 21.0
> 457505/teuthology.log:2026-08-11T16:16:57.587 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 22.0, inc_sync_duration1 - 9.0
> 457505/teuthology.log:2026-08-11T16:17:37.341 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 22.0, inc_sync_duration2 - 20.0
> 457506/teuthology.log:2026-08-11T16:16:09.330 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 23.0, inc_sync_duration1 - 11.0
> 457506/teuthology.log:2026-08-11T16:16:48.693 DEBUG:tasks.cephfs.test_mirroring:full_sync_duration - 23.0, inc_sync_duration2 - 21.0
> ```

This is effectively 2x the latency :/

**Comment by vshankar** [comment]
@kotreshhr - change seems reasonable. I will run this through tests to assess once more.

**Comment by vshankar** [comment]
jenkins test windows

**Comment by kotreshhr** [comment]
jenkins test windows

**Comment by kotreshhr** [comment]
jenkins test windows

**Comment by kotreshhr** [comment]
Rebased to fix jenkins windows test

**Comment by vshankar** [comment]
This PR has been added to [main integration testing by vshankar started 2026-09-02-01:48](https://tracker.ceph.com/issues/80177).

**Comment by vshankar** [comment]
jenkins test windows

**Comment by vshankar** [comment]
@kotreshhr - I haven't looked closely at the results, but fs:mirror ojbs failed with

> evicting unresponsive client trial141:mirror

See https://pulpito.ceph.com/vshankar-2026-09-02_05:42:23-fs:mirror-wip-vshankar-testing-20260902.014800-distro-default-trial/

Mind having a look?

**Comment by kotreshhr** [comment]
> @kotreshhr - I haven't looked closely at the results, but fs:mirror ojbs failed with
> 
> > evicting unresponsive client trial141:mirror
> 
> See https://pulpito.ceph.com/vshankar-2026-09-02_05:42:23-fs:mirror-wip-vshankar-testing-20260902.014800-distro-default-trial/
> 
> Mind having a look?

@vshankar 
The warning is expected because of the fix and test introduced `test_cephfs_mirror_sigkill_during_sync_recovers` as part of PR https://github.com/ceph/ceph/pull/71319 The commit message explicitly mentions this

```
...As a side effect, the new mirror can acquire the flock only after the MDS
clears the lock held by the previous session. With the default session_autoclose
of 300 seconds, a mirror daemon that restarts after a crash may wait roughly
five minutes before it can recover if the killed process died while holding
the flock.
```

Let's confirm the above test generated

1. The warning is generated at "2026-09-02T06:56:07.925454+0000" as per below warning
      "2026-09-02T06:56:07.925454+0000 mds.c (mds.0) 1 : cluster [WRN] evicting unresponsive client trial141:mirror (13979), after 316.4 seconds" in cluster log

2. The teuthology test which was running during test is as below.
     ```
    2026-09-02T06:49:36.398 DEBUG:teuthology.orchestra.run.trial141:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph log 'Starting test tasks.cephfs.test_mirroring.TestMirroring.test_cephfs_mirror_sigkill_during_sync_recovers'
    2026-09-02T06:56:07.101 ...
    2026-09-02T06:56:07.101 ...
    ...
    ...
    2026-09-02T06:58:09.061 DEBUG:teuthology.orchestra.run.trial141:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph log 'Ended test tasks.cephfs.test_mirroring.TestMirroring.test_cephfs_mirror_sigkill_during_sync_recovers'

     ```
This confirms that the test `test_cephfs_mirror_sigkill_during_sync_recovers` generated the warning

The warning should be added ignore list in mirroring test suite. But I am not sure why the test run with PR 71319 didn't produce it here https://pulpito.ceph.com/khiremat-2026-08-27_03:00:29-fs-wip-khiremat-testing-20260826.014416-distro-default-trial/

**Comment by kotreshhr** [comment]
> > @kotreshhr - I haven't looked closely at the results, but fs:mirror ojbs failed with
> > > evicting unresponsive client trial141:mirror
> > 
> > 
> > See https://pulpito.ceph.com/vshankar-2026-09-02_05:42:23-fs:mirror-wip-vshankar-testing-20260902.014800-distro-default-trial/
> > Mind having a look?
> 
> @vshankar The warning is expected because of the fix and test introduced `test_cephfs_mirror_sigkill_during_sync_recovers` as part of PR #71319 The commit message explicitly mentions this
> 
> ```
> ...As a side effect, the new mirror can acquire the flock only after the MDS
> clears the lock held by the previous session. With the default session_autoclose
> of 300 seconds, a mirror daemon that restarts after a crash may wait roughly
> five minutes before it can recover if the killed process died while holding
> the flock.
> ```
> 
> Let's confirm the above test generated
> 
> 1. The warning is generated at "2026-09-02T06:56:07.925454+0000" as per below warning
>    "2026-09-02T06:56:07.925454+0000 mds.c (mds.0) 1 : cluster [WRN] evicting unresponsive client trial141:mirror (13979), after 316.4 seconds" in cluster log
> 2. The teuthology test which was running during test is as below.
>    ```
>    2026-09-02T06:49:36.398 DEBUG:teuthology.orchestra.run.trial141:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph log 'Starting test tasks.cephfs.test_mirroring.TestMirroring.test_cephfs_mirror_sigkill_during_sync_recovers'
>    2026-09-02T06:56:07.101 ...
>    2026-09-02T06:56:07.101 ...
>    ...
>    ...
>    2026-09-02T06:58:09.061 DEBUG:teuthology.orchestra.run.trial141:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph log 'Ended test tasks.cephfs.test_mirroring.TestMirroring.test_cephfs_mirror_sigkill_during_sync_recovers'
>    ```
> 
> This confirms that the test `test_cephfs_mirror_sigkill_during_sync_recovers` generated the warning
> 
> The warning should be added ignore list in mirroring test suite. But I am not sure why the test run with PR 71319 didn't produce it here https://pulpito.ceph.com/khiremat-2026-08-27_03:00:29-fs-wip-khiremat-testing-20260826.014416-distro-default-trial/

Refreshed the PR adding the warning 'evicting unresponsive client' to ignore list in fs:mirror
@vshankar ptal

**Comment by vshankar** [comment]
> Refreshed the PR adding the warning 'evicting unresponsive client' to ignore list in fs:mirror

Sounds good. Will retest.

**Comment by kotreshhr** [comment]
jenkins test windows

**Comment by kotreshhr** [comment]
Rebased again to see if windows ci test passes with rebase

**Comment by kotreshhr** [comment]
Refreshed the PR by adding snap flush pending check for mtime matching case as well to avoid taking rdlock
https://github.com/ceph/ceph/pull/69895/changes/46359e1307e679152b792169149c3080fe0101fa#diff-277dc6e796ccecb6aa14c9357f7a86898d6ddcf4113c110a4e00e0a10f6fefa7R12700

**Comment by kotreshhr** [comment]
jenkins test make check

**Comment by kotreshhr** [comment]
jenkins test windows

**Comment by kotreshhr** [comment]
make check and windows ci job failed because of 
unittest_librbd crash in TestIOWithIOHint
ceph_test_librbd.TestMigration.SnapsSnapExpandAfterPrepare

Not related to snapdiff libcephfs test as earlier

**Comment by vshankar** [comment]
> Refreshed the PR by adding snap flush pending check for mtime matching case as well to avoid taking rdlock

Sounds good. Will run through tests.

**Comment by kotreshhr** [comment]
Rebased because of merge conflict

**Comment by kotreshhr** [comment]
jenkins test windows

**Comment by kotreshhr** [comment]
jenkins test make check

**Comment by kotreshhr** [comment]
jenkins test api

**Review by vshankar — APPROVED** [review]

**Comment by kotreshhr** [comment]
@vshankar I think the merge conflict resolution needs to be done carefully. I will take one more look. please don't merge

**Comment by kotreshhr** [comment]
> @vshankar I think the merge conflict resolution needs to be done carefully. I will take one more look. please don't merge

@vshankar I have carefully done the conflict resolution so that both the multiversion case and snap flush pending case is handled. pedal

**Comment by kotreshhr** [comment]
Sample mirror qa run with the rebased changes - 
https://pulpito-ng.ceph.com/runs/khiremat-2026-09-08_04:55:11-fs-wip-khiremat-69895-snapdiff-08092026-distro-default-trial

@vshankar The run looks good but required a warning 'evicting unresponsive clients' to be added to ignore list of valgrind mirror suite as well. I was only added to run without valgrind. I will refresh the PR adding that.

**Comment by kotreshhr** [comment]
> Sample mirror qa run with the rebased changes - https://pulpito-ng.ceph.com/runs/khiremat-2026-09-08_04:55:11-fs-wip-khiremat-69895-snapdiff-08092026-distro-default-trial
> 
> @vshankar The run looks good but required a warning 'evicting unresponsive clients' to be added to ignore list of valgrind mirror suite as well. I was only added to run without valgrind. I will refresh the PR adding that.

Done - https://github.com/ceph/ceph/pull/69895/changes/5489bf408ae0b851fb0cccefb683a9429da23679#diff-893142109427b64c905a19b784351bef3cee9c627408245ef716bd7d0acc62dfR35

**Comment by vshankar** [comment]
This change is good to merge as the mirror dameon tests have passed. See: https://pulpito.ceph.com/vshankar-2026-09-18_14:09:51-fs-wip-vshankar-testing-20260916.092455-distro-default-trial/

However, there is still a corner case in snapdiff that could cause it to miss reporting file. See: https://tracker.ceph.com/issues/80634

I have a fix for the above tracker and it's based on top of this change.
